src/Reference.php

```diff
@@ -0,0 +1,237 @@
+     *
+     * @param array $defaults Properties
+     *
+     * @return Model
+     */
+    public function getModel($defaults = [])
+    {
+        // set table_alias
+        if (!isset($defaults['table_alias'])) {
+            if (!$this->table_alias) {
+                $this->table_alias = $this->link;
+                $this->table_alias = preg_replace('/_id/', '', $this->table_alias);
+                $this->table_alias = preg_replace('/([a-zA-Z])[a-zA-Z]*[^a-zA-Z]*/', '\1', $this->table_alias);
+                if (isset($this->owner->table_alias)) {
+                    $this->table_alias = $this->owner->table_alias.'_'.$this->table_alias;
+                }
+            }
+            $defaults['table_alias'] = $this->table_alias;
+        }
+
+        // if model is Closure, then call it and return model
+        if (is_object($this->model) && $this->model instanceof \Closure) {
+            $c = $this->model;
+
+            $c = $c($this->owner, $this, $defaults);
+            if (!$c->persistence && $this->owner->persistence) {
+                $c = $this->owner->persistence->add($c, $defaults);
+            }
+
+            return $c;
+        }
+
+        // if model is set, then return clone of this model
+        if (is_object($this->model)) {
+            $c = clone $this->model;
+            if (!$this->model->persistence && $this->owner->persistence) {
+                $this->owner->persistence->add($c, $defaults);
+            }
+
+            return $c;
+        }
+
+        // last effort - try to add model
+        if (is_array($this->model)) {
+            $model = $this->model[0];
+            $md = $this->model;
+            unset($md[0]);
+
+            $defaults = array_merge($md, $defaults);
+        } else {
+            $model = $this->model;
+        }
+
+        $p = $this->owner->persistence;
+
+        return $p->add($p->normalizeClassName($model, 'Model'), $defaults);
+    }
+
+    /**
+     * Returns referenced model without any extra conditions.
+     *
+     * @param array $defaults Properties
+     *
+     * @return Model
+     */
+    public function ref($defaults = [])
+    {
+        return $this->getModel($defaults);
+    }
+
+    /**
+     * If you have set $their_field property correctly, then calling this method
+     * will traverse into a related model and fetch the type of their field. This
+     * method will be extra careful to avoid any loops.
+     */
+    public function guessFieldType($their_field = null)
+    {
+        if (!is_scalar($their_field)) {
+            return;
+        }
+
+        // We shouldn't traverse into ourselves recursively
+        $cl = get_class($this->owner).'::'.$this->short_name;
+
+        // see where we shouldn't traverse
+        if (isset($this->owner->_do_not_traverse_into)) {
+            $dnti = $this->owner->_do_not_traverse_into;
+
+            if (isset($dnti[$cl])) {
+                return;
+            }
+        }
+        $dnti[$cl] = true;
```

Comment:
$dnti is undefined here if $this->owner->_do_not_traverse_into is not defined at all in model